### PR TITLE
nixos-rebuild: Support `system.nix` by default

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

(in addition to various minor improvements)

Previously, `nixos-rebuild -A foo` would build the `default.nix` in the current directory, while `nixos-rebuild` without `-A` would not use file-based building at all.

This difference compared to what `nix-build` would do is confusing. Furthermore, using `default.nix` is not great, because there's many of those files all around, and they're almost never for NixOS.

This PR fixes those two issues by:
- Removing support for using `default.nix` as a default when using `--attr`, **this is a breaking change**, but support for this has not made it to a stable branch
- Using the `<nixos-system>` entry in NIX_PATH as the default if it exists
- Otherwise, use the `system.nix` file in the current directory or any of its parents, if it exists, as a default
- Otherwise, use `/etc/nixos/system.nix` as a default

While `system.nix` could be conflated with the `system` Nixpkgs argument, there's no practical way for this to cause problems, because `system` is never specified in a file.

Having this finally allows using `nixos-rebuild` without any arguments and without any `NIX_PATH` entries, effectively a way to declaratively use NixOS without `nix-channel`!

I think it's important to get this for the next release, so that we don't bake in stability for the `default.nix` behavior.

Unfortunately I'm a bit low on time, but if @amozeo (who pioneered the file-based building with https://github.com/NixOS/nixpkgs/pull/320462) could take this PR to its conclusion I'd be very grateful!

## Things done
- [ ] Tests
- [ ] Documentation

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by amozeo** [comment]
hey @infinisil, I wrote tests for configuration detection. I think you might be interested looking into it
it is available here: [permalink](https://github.com/amozeo/nixpkgs/tree/734b943e8261c7eb453f771ec479685c19bd5d3b), [latest](https://github.com/amozeo/nixpkgs/tree/pkgs/nixos-rebuild/test-configuration-detection), [permalink diff](https://github.com/infinisil/nixpkgs/compare/nixos-rebuild-improvements...amozeo:nixpkgs:734b943e8261c7eb453f771ec479685c19bd5d3b), [latest diff](https://github.com/infinisil/nixpkgs/compare/nixos-rebuild-improvements...amozeo:nixpkgs:pkgs/nixos-rebuild/test-configuration-detection)

**Comment by infinisil** [comment]
@amozeo Neat idea, it seems to work well with `system.nix`! Were you thinking of merging that idea into this PR? I'd prefer having smaller separate PRs that only focus on fixing one problem :)

**Comment by amozeo** [comment]
@infinisil how I should split it then?

**Comment by infinisil** [comment]
Honestly https://github.com/NixOS/nixpkgs/pull/327740 seemed fine to me. I guess it would be best to have this PR first though. I'm unfortunately a bit low on time right now, so I'd appreciate if you could help out!

In particular I think this PR should get done before the release in ~2 months, because I don't think the current behavior regarding `default.nix` is great, and we'd be locking that into stable :sweat:

**Comment by nyabinary** [comment]
We should probably get this moving if we want to make it before next release right?

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh [hunk 1]
```diff
@@ -368,11 +368,44 @@ fi
...
+# Default build file if none is specified in the arguments
+if [[ -z $buildFile ]]; then
+    # From NIX_PATH
+    if ! buildFile=$(runCmd nix-instantiate --find-file nixos-system 2>/dev/null); then
+        # From system.nix up from the current directory
+        if ! buildFile=$(findInParents "$PWD" system.nix); then
+            # Hardcoded to /etc/nixos/system.nix
+            if [[ -f "/etc/nixos/system.nix" ]]; then
+                buildFile=/etc/nixos/system.nix
+            fi
```
Comment (amozeo) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:399]: I don't think this path should be hardcoded in nixos-rebuild. Setup with channels doesn't have /etc/nixos/configuration.nix hardcoded and uses nixos-config in `NIX_PATH` instead to find its configuration there. If the /etc/nixos/system.nix must be hardcoded, it should be in NixOS modules instead, similar to channel setup.

Comment (amozeo) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:399]: flake path `/etc/nixos/flake.nix` is hardcoded in `nixos-rebuild`, I don't think the issue that I brough up has any significance anymore

**Comment by amozeo** [comment]
@infinisil I have force-updated the branch on my fork that I linked before.
[permalink](https://github.com/amozeo/nixpkgs/tree/ec906113ced73d2368edeb3e96b8fd58cdd89225), [latest](https://github.com/amozeo/nixpkgs/tree/pkgs/nixos-rebuild/test-configuration-detection), [permalink diff](https://github.com/nixos/nixpkgs/compare/master...ec906113ced73d2368edeb3e96b8fd58cdd89225), [latest diff](https://github.com/infinisil/nixpkgs/compare/nixos-rebuild-improvements...amozeo:nixpkgs:pkgs/nixos-rebuild/test-configuration-detection).
I want to make sure before I do something. Are you sure that I should make 4 separate pull requests for each feature that this batch of commits brings?
These pull requests would be named:
- nixos-{rebuild, install}: Support `system.nix` by default and improve configuration detection
- lib/version-info-fixup: create
- nixos-generate-config: Introduce templates
- nixos-rebuild: Add test for configuration detection

**Comment by nyabinary** [comment]
> @infinisil I have force-updated the branch on my fork that I linked before. [permalink](https://github.com/amozeo/nixpkgs/tree/ec906113ced73d2368edeb3e96b8fd58cdd89225), [latest](https://github.com/amozeo/nixpkgs/tree/pkgs/nixos-rebuild/test-configuration-detection), [permalink diff](https://github.com/nixos/nixpkgs/compare/master...ec906113ced73d2368edeb3e96b8fd58cdd89225), [latest diff](https://github.com/infinisil/nixpkgs/compare/nixos-rebuild-improvements...amozeo:nixpkgs:pkgs/nixos-rebuild/test-configuration-detection). I want to make sure before I do something. Are you sure that I should make 4 separate pull requests for each feature that this batch of commits brings? These pull requests would be named:
> 
>     * nixos-{rebuild, install}: Support `system.nix` by default and improve configuration detection
> 
>     * lib/version-info-fixup: create
> 
>     * nixos-generate-config: Introduce templates
> 
>     * nixos-rebuild: Add test for configuration detection

Doesn't hurt to open the PR right now as all of them together, and then you can split them later if required.

**Comment by nyabinary** [comment]
Bump :P 24.11 is rapidly approaching

**Comment by emilazy** [comment]
We’re in feature freeze, so this would need @RossComputerGuy’s sign‐off even if it wasn’t a draft.

**Comment by RossComputerGuy** [comment]
Yeah, let's kill the default.nix post release. This should be fine.

**Comment by nyabinary** [comment]
> Yeah, let's kill the default.nix post release. This should be fine.

bump :P

**Comment by justinas** [comment]
My 2c as a big fan of "flake-less, channel-less" model: this is a great change.

Would we like to default to building `/etc/nixos/system.nix` if it exists, over building `configuration.nix`, similarly how it is [done for flakes](https://github.com/NixOS/nixpkgs/blob/1908ad1385b4dae62d7c2225d25a39006a9177bc/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L440-L444)? This would pretty much non-breaking if the priority remains: build `flake.nix` if it exists, build `system.nix` if it exists, else fall back on `configuration.nix` + `<nixpkgs/nixos>`. This approach could still break for people who happen to be using `/etc/nixos/configuration.nix`, *and* have an unrelated file named `/etc/nixos/system.nix`. Should be fine if documented in breaking changes?

Are there any plans to make  `system.nix` and the plumbing such as `lib/eval-config.nix` more of first-class citizens? Right now there's quite a big disconnect in how NixOS is built using flakes (explicit call to `lib.nixosSystem` - another variation of basically `lib/eval-config.nix`), and how it is usually built in a flake-less way (`/etc/nixos/configuration.nix` and `<nixpkgs/nixos>` inferred from the environment, these are sort of "implementation details" that users often remain ignorant of).

In other words, do we see a future where perhaps `nixos-generate-config` also generates a `/etc/nixos/system.nix` with something like:

```nix
let
  nixpkgs = <nixpkgs> # keeping compat with the "channelful" model by default
in
import "${nixpkgs}/nixos/lib/eval-config.nix" {
  system = null;
  modules = [
    ./configuration.nix
  ];
}
```

and this becomes a supported/recommended approach?

Let me know if this issue is not a place to discuss this, I'll make a new one. I'm just curious and excited about any ways to demystify the NixOS build process and decouple it somewhat from `<nixpkgs/nixos>` / `<nixos-config>` assumptions / implementation details in `nixos-rebuild`.

**Comment by amozeo** [comment]
I created https://github.com/NixOS/nixpkgs/pull/389487 that addresses issues pointed out in this pull request.
_For anyone subscribed to this pull request_

**Comment by bjornfor** [comment]
I guess this can be closed now that https://github.com/NixOS/nixpkgs/pull/493229 has been merged? (Btw, thanks to those who worked on this feature!)
